File: genium/src/main/java/com/here/genium/generator/cpp/CppTypeMapper.java
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp;

import com.here.genium.cli.GeniumExecutionException;
import com.here.genium.model.common.InstanceRules;
import com.here.genium.model.cpp.*;
import java.util.Collections;
import org.franca.core.franca.*;

/**
 * Maps Franca type references to their C++ counterparts. These references are used as parameters,
 * in typedefs, array members etc.
 */
public class CppTypeMapper {

  private static final CppTypeRef BASIC_STRING_CHAR_TYPE =
      CppTemplateTypeRef.Companion.create(
          CppTemplateTypeRef.TemplateClass.BASIC_STRING, CppPrimitiveTypeRef.Companion.getCHAR());

  public static final CppTypeRef STD_ERROR_CODE_TYPE =
      new CppComplexTypeRef(
          "::std::error_code", Collections.singletonList(CppLibraryIncludes.SYSTEM_ERROR));

  public static final CppTypeRef STRING_TYPE =
      new CppTypeDefRef("::std::string", BASIC_STRING_CHAR_TYPE.includes, BASIC_STRING_CHAR_TYPE);

  private static final String ENUM_HASH_CLASS_NAME = "EnumHash";

  private final CppIncludeResolver includeResolver;
  private final CppNameResolver nameResolver;
  private final String internalNamespace;

  public CppTypeMapper(
      final CppIncludeResolver includeResolver,
      final CppNameResolver nameResolver,
      final String internalNamespace) {
    this.includeResolver = includeResolver;
    this.nameResolver = nameResolver;
    this.internalNamespace = internalNamespace;
  }

  public CppTypeRef getEnumHashType() {
    String name = CppNameRules.joinFullyQualifiedName(internalNamespace, ENUM_HASH_CLASS_NAME);
    return new CppComplexTypeRef(name, Collections.singletonList(CppLibraryIncludes.ENUM_HASH));
  }

  public CppTypeRef getReturnWrapperType(final CppTypeRef outArgType, final CppTypeRef errorType) {
    return CppTemplateTypeRef.Companion.create(
        internalNamespace, CppTemplateTypeRef.TemplateClass.RETURN, outArgType, errorType);
  }

  public CppTypeRef getByteBufferType() {
    CppTypeRef byteBufferType =
        CppTemplateTypeRef.Companion.create(
            CppTemplateTypeRef.TemplateClass.VECTOR, CppPrimitiveTypeRef.Companion.getUINT8());
    return CppTemplateTypeRef.Companion.create(
        CppTemplateTypeRef.TemplateClass.SHARED_POINTER, byteBufferType);
  }

  public CppTypeRef map(FTypeRef type) {

    return type.getDerived() != null ? mapDerived(type.getDerived()) : mapPredefined(type);
  }

  private CppTypeRef mapDerived(FType derived) {
    if (derived instanceof FTypeDef) {
      return mapTypeDef((FTypeDef) derived);
    }
    if (derived instanceof FArrayType) {
      return mapArray((FArrayType) derived);
    }
    if (derived instanceof FMapType) {
      return mapMapType((FMapType) derived);
    }
    if (derived instanceof FCompoundType || derived instanceof FEnumerationType) {
      return mapComplexType(derived);
    }
    throw new GeniumExecutionException("Unmapped derived type: " + derived.getName());
  }

  private CppTypeRef mapTypeDef(final FTypeDef francaTypeDef) {

    if (InstanceRules.isInstanceId(francaTypeDef)) {
      String fullyQualifiedName =
          nameResolver.getFullyQualifiedName((FInterface) francaTypeDef.eContainer());
      CppComplexTypeRef instanceType = new CppInstanceTypeRef(fullyQualifiedName);
      instanceType.includes.add(includeResolver.resolveInclude(francaTypeDef));

      return CppTemplateTypeRef.Companion.create(
          CppTemplateTypeRef.TemplateClass.SHARED_POINTER, instanceType);
    } else {

      String fullyQualifiedName = nameResolver.getFullyQualifiedName(francaTypeDef);
      CppTypeRef actualType = map(francaTypeDef.getActualType());

      return new CppTypeDefRef(
          fullyQualifiedName,
          Collections.singletonList(includeResolver.resolveInclude(francaTypeDef)),
          actualType);
    }
  }

  private CppTypeRef mapArray(final FArrayType francaArrayType) {

    String fullyQualifiedName = nameResolver.getFullyQualifiedName(francaArrayType);
    CppTypeRef elementType = map(francaArrayType.getElementType());
    CppTypeRef arrayType =
        CppTemplateTypeRef.Companion.create(CppTemplateTypeRef.TemplateClass.VECTOR, elementType);

    return new CppTypeDefRef(
        fullyQualifiedName,
        Collections.singletonList(includeResolver.resolveInclude(francaArrayType)),
        arrayType);
  }

  private CppTypeRef mapMapType(final FMapType francaMapType) {

    String fullyQualifiedName = nameResolver.getFullyQualifiedName(francaMapType);
    CppTypeRef mapType =
        wrapMap(map(francaMapType.getKeyType()), map(francaMapType.getValueType()));

    return new CppTypeDefRef(
        fullyQualifiedName,
        Collections.singletonList(includeResolver.resolveInclude(francaMapType)),
        mapType);
  }

  public CppTypeRef wrapMap(final CppTypeRef key, final CppTypeRef value) {
    if (key.getRefersToEnumType()) {
      return CppTemplateTypeRef.Companion.create(
          CppTemplateTypeRef.TemplateClass.MAP, key, value, getEnumHashType());
    } else {
      return CppTemplateTypeRef.Companion.create(CppTemplateTypeRef.TemplateClass.MAP, key, value);
    }
  }

  public CppTypeRef mapComplexType(final FModelElement francaElement) {

    return new CppComplexTypeRef(
        nameResolver.getFullyQualifiedName(francaElement),
        Collections.singletonList(includeResolver.resolveInclude(francaElement)),
        francaElement instanceof FEnumerationType);
  }

  private CppTypeRef mapPredefined(final FTypeRef type) {

    switch (type.getPredefined().getValue()) {
      case FBasicTypeId.BOOLEAN_VALUE:
        return CppPrimitiveTypeRef.Companion.getBOOL();

      case FBasicTypeId.FLOAT_VALUE:
        return CppPrimitiveTypeRef.Companion.getFLOAT();

      case FBasicTypeId.DOUBLE_VALUE:
        return CppPrimitiveTypeRef.Companion.getDOUBLE();

      case FBasicTypeId.INT8_VALUE:
        return CppPrimitiveTypeRef.Companion.getINT8();

      case FBasicTypeId.INT16_VALUE:
        return CppPrimitiveTypeRef.Companion.getINT16();

      case FBasicTypeId.INT32_VALUE:
        return CppPrimitiveTypeRef.Companion.getINT32();

      case FBasicTypeId.INT64_VALUE:
        return CppPrimitiveTypeRef.Companion.getINT64();

      case FBasicTypeId.UINT8_VALUE:
        return CppPrimitiveTypeRef.Companion.getUINT8();

      case FBasicTypeId.UINT16_VALUE:
        return CppPrimitiveTypeRef.Companion.getUINT16();

      case FBasicTypeId.UINT32_VALUE:
        return CppPrimitiveTypeRef.Companion.getUINT32();

      case FBasicTypeId.UINT64_VALUE:
        return CppPrimitiveTypeRef.Companion.getUINT64();

      case FBasicTypeId.STRING_VALUE:
        return STRING_TYPE;

      case FBasicTypeId.BYTE_BUFFER_VALUE:
        return getByteBufferType();

      case FBasicTypeId.UNDEFINED_VALUE:
        return CppPrimitiveTypeRef.Companion.getVOID();

      default:
        throw new GeniumExecutionException(
            "unmapped predefined [" + type.getPredefined().getName() + "]");
    }
  }
}

File: genium/src/main/java/com/here/genium/generator/cpp/CppTypeMapper.kt
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */

package com.here.genium.generator.cpp

import com.here.genium.cli.GeniumExecutionException
import com.here.genium.model.cpp.CppComplexTypeRef
import com.here.genium.model.cpp.CppPrimitiveTypeRef
import com.here.genium.model.cpp.CppTemplateTypeRef
import com.here.genium.model.cpp.CppTemplateTypeRef.TemplateClass
import com.here.genium.model.cpp.CppTypeDefRef
import com.here.genium.model.cpp.CppTypeRef
import com.here.genium.model.lime.LimeArray
import com.here.genium.model.lime.LimeAttributeType.CPP
import com.here.genium.model.lime.LimeAttributeValueType.EXTERNAL_TYPE
import com.here.genium.model.lime.LimeBasicType
import com.here.genium.model.lime.LimeBasicType.TypeId
import com.here.genium.model.lime.LimeContainer
import com.here.genium.model.lime.LimeEnumeration
import com.here.genium.model.lime.LimeException
import com.here.genium.model.lime.LimeMap
import com.here.genium.model.lime.LimeSet
import com.here.genium.model.lime.LimeStruct
import com.here.genium.model.lime.LimeType
import com.here.genium.model.lime.LimeTypeDef
import com.here.genium.model.lime.LimeTypeHelper
import com.here.genium.model.lime.LimeTypeRef

class CppTypeMapper(
    private val nameResolver: CppNameResolver,
    private val includeResolver: CppIncludeResolver,
    private val internalNamespace: List<String>
) {
    fun getReturnWrapperType(outArgType: CppTypeRef, errorType: CppTypeRef): CppTypeRef =
        CppTemplateTypeRef(
            TemplateClass.RETURN,
            outArgType,
            errorType,
            namespace = CppNameRules.joinFullyQualifiedName(internalNamespace)
        )

    fun mapType(limeTypeRef: LimeTypeRef): CppTypeRef {
        val result = mapType(limeTypeRef.type)
        val needsOptionalType = limeTypeRef.isNullable && limeTypeRef.type !is LimeContainer
        return if (needsOptionalType) createOptionalType(result) else result
    }

    fun mapInstanceType(
        limeContainer: LimeContainer,
        needsForwardDeclaration: Boolean
    ) = CppComplexTypeRef(
        fullyQualifiedName = nameResolver.getFullyQualifiedName(limeContainer),
        includes = includeResolver.resolveIncludes(limeContainer),
        needsForwardDeclaration = needsForwardDeclaration
    )

    private fun createOptionalType(cppTypeRef: CppTypeRef) =
        CppTemplateTypeRef(
            TemplateClass.OPTIONAL,
            cppTypeRef,
            namespace = CppNameRules.joinFullyQualifiedName(internalNamespace)
        )

    private fun mapType(limeType: LimeType): CppTypeRef =
        when (limeType) {
            is LimeBasicType -> mapPredefined(limeType)
            is LimeTypeDef -> CppTypeDefRef(
                nameResolver.getFullyQualifiedName(limeType),
                includeResolver.resolveIncludes(limeType),
                mapType(limeType.typeRef)
            )
            is LimeStruct -> CppComplexTypeRef(
                nameResolver.getFullyQualifiedName(limeType),
                includeResolver.resolveIncludes(limeType)
            )
            is LimeEnumeration -> CppComplexTypeRef(
                nameResolver.getFullyQualifiedName(limeType),
                includeResolver.resolveIncludes(limeType),
                refersToValueType = true
            )
            is LimeArray ->
                CppTemplateTypeRef(TemplateClass.VECTOR, mapType(limeType.elementType))
            is LimeMap -> {
                val keyType = mapType(limeType.keyType)
                if (hasStdHash(limeType.keyType)) {
                    CppTemplateTypeRef(
                        TemplateClass.MAP,
                        keyType,
                        mapType(limeType.valueType)
                    )
                } else {
                    val hashType = CppTemplateTypeRef(
                        TemplateClass.HASH,
                        keyType,
                        namespace = internalNamespace.joinToString("::")
                    )
                    CppTemplateTypeRef(
                        TemplateClass.MAP,
                        keyType,
                        mapType(limeType.valueType),
                        hashType
                    )
                }
            }
            is LimeContainer -> {
                val instanceType =
                    mapInstanceType(limeType, !limeType.attributes.have(CPP, EXTERNAL_TYPE))
                CppTemplateTypeRef(TemplateClass.SHARED_POINTER, instanceType)
            }
            is LimeSet -> {
                val elementType = mapType(limeType.elementType)
                if (hasStdHash(limeType.elementType)) {
                    CppTemplateTypeRef(TemplateClass.SET, elementType)
                } else {
                    val hashType = CppTemplateTypeRef(
                        TemplateClass.HASH,
                        elementType,
                        namespace = internalNamespace.joinToString("::")
                    )
                    CppTemplateTypeRef(TemplateClass.SET, elementType, hashType)
                }
            }
            is LimeException -> STD_ERROR_CODE_TYPE
            else -> throw GeniumExecutionException("Unmapped type: " + limeType.name)
        }

    private fun mapPredefined(limeBasicType: LimeBasicType) =
        when (limeBasicType.typeId) {
            TypeId.VOID -> CppPrimitiveTypeRef.VOID
            TypeId.INT8 -> CppPrimitiveTypeRef.INT8
            TypeId.UINT8 -> CppPrimitiveTypeRef.UINT8
            TypeId.INT16 -> CppPrimitiveTypeRef.INT16
            TypeId.UINT16 -> CppPrimitiveTypeRef.UINT16
            TypeId.INT32 -> CppPrimitiveTypeRef.INT32
            TypeId.UINT32 -> CppPrimitiveTypeRef.UINT32
            TypeId.INT64 -> CppPrimitiveTypeRef.INT64
            TypeId.UINT64 -> CppPrimitiveTypeRef.UINT64
            TypeId.BOOLEAN -> CppPrimitiveTypeRef.BOOL
            TypeId.FLOAT -> CppPrimitiveTypeRef.FLOAT
            TypeId.DOUBLE -> CppPrimitiveTypeRef.DOUBLE
            TypeId.STRING -> STRING_TYPE
            TypeId.BLOB -> BYTE_BUFFER_POINTER_TYPE
            TypeId.DATE -> DATE_TYPE
        }

    companion object {
        val STD_ERROR_CODE_TYPE: CppTypeRef =
            CppComplexTypeRef("::std::error_code", listOf(CppLibraryIncludes.SYSTEM_ERROR))
        private val BASIC_STRING_CHAR_TYPE =
            CppTemplateTypeRef(TemplateClass.BASIC_STRING, CppPrimitiveTypeRef.CHAR)
        internal val STRING_TYPE =
            CppTypeDefRef("::std::string", BASIC_STRING_CHAR_TYPE.includes, BASIC_STRING_CHAR_TYPE)
        private val BYTE_BUFFER_ARRAY_TYPE =
            CppTemplateTypeRef(TemplateClass.VECTOR, CppPrimitiveTypeRef.UINT8)
        val BYTE_BUFFER_POINTER_TYPE =
            CppTemplateTypeRef(TemplateClass.SHARED_POINTER, BYTE_BUFFER_ARRAY_TYPE)
        private val DATE_TYPE =
            CppComplexTypeRef(
                "::std::chrono::system_clock::time_point",
                listOf(CppLibraryIncludes.CHRONO)
            )

        fun hasStdHash(limeType: LimeTypeRef): Boolean {
            val actualType =
                LimeTypeHelper.getActualType(limeType.type) as? LimeBasicType ?: return false
            return actualType.typeId != TypeId.BLOB && actualType.typeId != TypeId.DATE
        }
    }
}
